ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/DropSourceCommand.java

```diff
@@ -58,16 +71,60 @@ public DdlCommandResult run(MetaStore metaStore, boolean isValidatePhase) {
         new DropTopicCommand(dataSource.getKsqlTopic().getTopicName());
     metaStore.deleteSource(sourceName);
     dropTopicCommand.run(metaStore, isValidatePhase);
-    if (!isValidatePhase && dataSource.getKsqlTopic().getKsqlTopicSerDe().getSerDe()
-                            == DataSource.DataSourceSerDe.AVRO) {
+
+    deleteTopicIfNeeded(dataSource, isValidatePhase);
+
+    return new DdlCommandResult(true, "Source " + sourceName + " was dropped. "
+                                      + (withTopic ? "Topic '"
+                                                    + dataSource.getKsqlTopic().getTopicName()
+                                                    + "' was marked for deletion. Actual deletion "
+                                                    + "and removal from brokers may take some time "
+                                                    + "to complete." : ""));
+  }
+
+  private void deleteTopicIfNeeded(StructuredDataSource dataSource, boolean isValidatePhase) {
+    if (!isValidatePhase && withTopic) {
+
+      executeWithRetries(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          kafkaTopicClient.deleteTopics(
+              Collections.singletonList(dataSource.getKsqlTopic().getKafkaTopicName()));
+          return null;
+        }
+      }, "Could not delete the corresponding kafka topic: "
+           + dataSource.getKsqlTopic().getKafkaTopicName());
+
+      if (dataSource.getKsqlTopic().getKsqlTopicSerDe().getSerDe()
+          == DataSource.DataSourceSerDe.AVRO) {
+        executeWithRetries(new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            schemaRegistryClient
+                .deleteSubject(sourceName + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX);
+            return null;
+          }
+        }, "Could not clean up the schema registry for topic: " + sourceName);
+      }
+    }
+  }
+
+  private void executeWithRetries(Callable<Void> callable, String errorMessage) {
```

Comment:
Hmm. THis looks like a dup of https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java#L337

It might be better to just have a singular `deleteTopic` method in `KafkaTopicClient`, have that execute with retries, and call the singular version here. Would be a good simplification.

Reply:
One way to do this would be to try and generalize the `executeWithRetries` such that it can be used in all places, i.e., perhaps by making changing the function signature to:
`<T> T executeWithRetries(Callable<T> callable, String errorMessage)`

You can then maybe refactor into a single function that is called from both places

Reply:
This seems like a good change for a followup. I recommend capturing these details in a new issue and labeling them 'good first issue' and 'help wanted'. Then github will help new potential contributors find these issues and hence make useful and manageable first patches. See: https://help.github.com/articles/helping-new-contributors-find-your-project-with-labels/

Reply:
@apurvam those kinds of things should probably live in common and not be project-specific

Reply:
I like the @apurvam 's suggestion. This can be a good follow up change and can be used as a way to engage the community more!